### ISSUE TITLE
python not reloaded when code changed

### ISSUE THREAD

**Issue by fabricehong** [langchain-ai/langgraph-studio#84]
Hello,

When I change my code (exactly like the last video in https://blog.langchain.dev/langgraph-studio-the-first-agent-ide/), and do a re-run, the new code is not taken in account unless I restart (online button).

Is there something special I should do to make is work exactly like in the video ?

best

**Comment by dqbd** [comment]
Hello @fabricehong! Could you share the contents of the following command? 
```
docker info -f json
```

**Comment by fabricehong** [comment]
sure, here you are:
```
{
  "ID": "086e3677-3fba-42c6-b193-7fc364d444c8",
  "Containers": 3,
  "ContainersRunning": 3,
  "ContainersPaused": 0,
  "ContainersStopped": 0,
  "Images": 12,
  "Driver": "overlay2",
  "DriverStatus": [
    [
      "Backing Filesystem",
      "extfs"
    ],
    [
      "Supports d_type",
      "true"
    ],
    [
      "Using metacopy",
      "false"
    ],
    [
      "Native Overlay Diff",
      "true"
    ],
    [
      "userxattr",
      "false"
    ]
  ],
  "Plugins": {
    "Volume": [
      "local"
    ],
    "Network": [
      "bridge",
      "host",
      "ipvlan",
      "macvlan",
      "null",
      "overlay"
    ],
    "Authorization": null,
    "Log": [
      "awslogs",
      "fluentd",
      "gcplogs",
      "gelf",
      "journald",
      "json-file",
      "local",
      "logentries",
      "splunk",
      "syslog"
    ]
  },
  "MemoryLimit": true,
  "SwapLimit": true,
  "CpuCfsPeriod": true,
  "CpuCfsQuota": true,
  "CPUShares": true,
  "CPUSet": true,
  "PidsLimit": true,
  "IPv4Forwarding": true,
  "BridgeNfIptables": true,
  "BridgeNfIp6tables": true,
  "Debug": false,
  "NFd": 66,
  "OomKillDisable": false,
  "NGoroutines": 107,
  "SystemTime": "2024-08-16T18:28:48.044270552Z",
  "LoggingDriver": "json-file",
  "CgroupDriver": "cgroupfs",
  "CgroupVersion": "2",
  "NEventsListener": 13,
  "KernelVersion": "6.4.16-linuxkit",
  "OperatingSystem": "Docker Desktop",
  "OSVersion": "",
  "OSType": "linux",
  "Architecture": "aarch64",
  "IndexServerAddress": "https://index.docker.io/v1/",
  "RegistryConfig": {
    "AllowNondistributableArtifactsCIDRs": null,
    "AllowNondistributableArtifactsHostnames": null,
    "InsecureRegistryCIDRs": [
      "127.0.0.0/8"
    ],
    "IndexConfigs": {
      "docker.io": {
        "Name": "docker.io",
        "Mirrors": [],
        "Secure": true,
        "Official": true
      },
      "hubproxy.docker.internal:5555": {
        "Name": "hubproxy.docker.internal:5555",
        "Mirrors": [],
        "Secure": false,
        "Official": false
      }
    },
    "Mirrors": null
  },
  "NCPU": 10,
  "MemTotal": 8232673280,
  "GenericResources": null,
  "DockerRootDir": "/var/lib/docker",
  "HttpProxy": "http.docker.internal:3128",
  "HttpsProxy": "http.docker.internal:3128",
  "NoProxy": "hubproxy.docker.internal",
  "Name": "docker-desktop",
  "Labels": [],
  "ExperimentalBuild": false,
  "ServerVersion": "24.0.6",
  "Runtimes": {
    "io.containerd.runc.v2": {
      "path": "runc"
    },
    "runc": {
      "path": "runc"
    }
  },
  "DefaultRuntime": "runc",
  "Swarm": {
    "NodeID": "",
    "NodeAddr": "",
    "LocalNodeState": "inactive",
    "ControlAvailable": false,
    "Error": "",
    "RemoteManagers": null
  },
  "LiveRestoreEnabled": false,
  "Isolation": "",
  "InitBinary": "docker-init",
  "ContainerdCommit": {
    "ID": "8165feabfdfe38c65b599c4993d227328c231fca",
    "Expected": "8165feabfdfe38c65b599c4993d227328c231fca"
  },
  "RuncCommit": {
    "ID": "v1.1.8-0-g82f18fe",
    "Expected": "v1.1.8-0-g82f18fe"
  },
  "InitCommit": {
    "ID": "de40ad0",
    "Expected": "de40ad0"
  },
  "SecurityOptions": [
    "name=seccomp,profile=unconfined",
    "name=cgroupns"
  ],
  "Warnings": [
    "WARNING: daemon is not using the default seccomp profile"
  ],
  "ClientInfo": {
    "Debug": false,
    "Version": "24.0.6",
    "GitCommit": "ed223bc",
    "GoVersion": "go1.20.7",
    "Os": "darwin",
    "Arch": "arm64",
    "BuildTime": "Mon Sep  4 12:28:49 2023",
    "Context": "desktop-linux",
    "Plugins": [
      {
        "SchemaVersion": "0.1.0",
        "Vendor": "Docker Inc.",
        "Version": "v0.11.2-desktop.5",
        "ShortDescription": "Docker Buildx",
        "Name": "buildx",
        "Path": "/Users/hong.fabrice/.docker/cli-plugins/docker-buildx",
        "ShadowedPaths": [
          "/usr/local/lib/docker/cli-plugins/docker-buildx"
        ]
      },
      {
        "SchemaVersion": "0.1.0",
        "Vendor": "Docker Inc.",
        "Version": "v2.22.0-desktop.2",
        "ShortDescription": "Docker Compose",
        "Name": "compose",
        "Path": "/Users/hong.fabrice/.docker/cli-plugins/docker-compose",
        "ShadowedPaths": [
          "/usr/local/lib/docker/cli-plugins/docker-compose"
        ]
      },
      {
        "SchemaVersion": "0.1.0",
        "Vendor": "Docker Inc.",
        "Version": "v0.1.0",
        "ShortDescription": "Docker Dev Environments",
        "Name": "dev",
        "Path": "/Users/hong.fabrice/.docker/cli-plugins/docker-dev",
        "ShadowedPaths": [
          "/usr/local/lib/docker/cli-plugins/docker-dev"
        ]
      },
      {
        "SchemaVersion": "0.1.0",
        "Vendor": "Docker Inc.",
        "Version": "v0.2.20",
        "ShortDescription": "Manages Docker extensions",
        "Name": "extension",
        "Path": "/Users/hong.fabrice/.docker/cli-plugins/docker-extension",
        "ShadowedPaths": [
          "/usr/local/lib/docker/cli-plugins/docker-extension"
        ]
      },
      {
        "SchemaVersion": "0.1.0",
        "Vendor": "Docker Inc.",
        "Version": "v0.1.0-beta.8",
        "ShortDescription": "Creates Docker-related starter files for your project",
        "Name": "init",
        "Path": "/Users/hong.fabrice/.docker/cli-plugins/docker-init",
        "ShadowedPaths": [
          "/usr/local/lib/docker/cli-plugins/docker-init"
        ]
      },
      {
        "SchemaVersion": "0.1.0",
        "Vendor": "Anchore Inc.",
        "Version": "0.6.0",
        "ShortDescription": "View the packaged-based Software Bill Of Materials (SBOM) for an image",
        "URL": "https://github.com/docker/sbom-cli-plugin",
        "Name": "sbom",
        "Path": "/Users/hong.fabrice/.docker/cli-plugins/docker-sbom",
        "ShadowedPaths": [
          "/usr/local/lib/docker/cli-plugins/docker-sbom"
        ]
      },
      {
        "SchemaVersion": "0.1.0",
        "Vendor": "Docker Inc.",
        "Version": "v0.26.0",
        "ShortDescription": "Docker Scan",
        "Name": "scan",
        "Path": "/Users/hong.fabrice/.docker/cli-plugins/docker-scan",
        "ShadowedPaths": [
          "/usr/local/lib/docker/cli-plugins/docker-scan"
        ]
      },
      {
        "SchemaVersion": "0.1.0",
        "Vendor": "Docker Inc.",
        "Version": "v1.0.7",
        "ShortDescription": "Docker Scout",
        "Name": "scout",
        "Path": "/Users/hong.fabrice/.docker/cli-plugins/docker-scout",
        "ShadowedPaths": [
          "/usr/local/lib/docker/cli-plugins/docker-scout"
        ]
      }
    ],
    "Warnings": null
  }
}
```

**Comment by dqbd** [comment]
Judging by the provided `docker info -f json` payload, you need to upgrade to latest Docker Compose and/or Docker Desktop. 

We rely on Docker Compose Watch, which is available from 2.25.0, while you're currently running 2.22.0 at the moment. A toast should be already showing when opening the app, if that's not the case, feel free to reopen the issue!

**Comment by saitaiky** [comment]
Hi @dqbd , sorry for bothering you but I have the latest docker desktop with the latest docker compose version -  @"Version":"v2.32.4-desktop.1". Also I have installed 0.0.36 LangGraph Studio, but still the hot reload doesn't work.

When I make changes to the source code, it does detect the change, but the status at the bottom left never change to stop and back to online for the rebuild process. Do you know why? I have been fixing it for 2 days.

```
langgraph-api-1       | info | Postgres pool stats
langgraph-api-1       | info | Redis pool stats
langgraph-api-1       | info | Sweeped runs
langgraph-api-1       | info | Worker stats
                      ⦿ Syncing service "langgraph-api" after 1 changes were detected
langgraph-api-1       | info | Postgres pool stats
langgraph-api-1       | info | Redis pool stats
langgraph-api-1       | info | Worker stats
langgraph-api-1       | info | GET /threads/f1bbaa3c-cb9a-48b3-9aaa-9d2d
```
